### ISSUE TITLE
Run gauge tests in parallel with sahi

### ISSUE THREAD

**Issue by kupna** [getgauge/gauge-java#61]
Hi!

For performance issues I was trying to run the gauge tests with sahi in parallel.

I used the following command:

gauge
--parallel
--tags="setup&!inprogress&!ei&basicIndependent"
./specs

I was trying for my tests but I got the following console output:

12:08:30.447 [go] setting environment variable 'GO_SERVER_URL' to value 'https://satwehrlgo:8154/go/'
12:08:30.447 [go] setting environment variable 'GO_TRIGGER_USER' to value 'changes'
12:08:30.447 [go] setting environment variable 'GO_PIPELINE_NAME' to value 'engage-gauge-reference-data'
12:08:30.447 [go] setting environment variable 'GO_PIPELINE_COUNTER' to value '48'
12:08:30.447 [go] setting environment variable 'GO_PIPELINE_LABEL' to value '2016.2.EAP.10027'
12:08:30.447 [go] setting environment variable 'GO_STAGE_NAME' to value 'acceptance-referencedata'
12:08:30.447 [go] setting environment variable 'GO_STAGE_COUNTER' to value '1'
12:08:30.447 [go] setting environment variable 'GO_JOB_NAME' to value 'referencedataIndependent'
12:08:30.447 [go] setting environment variable 'GO_REVISION_ENGAGE_ACCEPTANCE_SAHI' to value '40610'
12:08:30.447 [go] setting environment variable 'GO_TO_REVISION_ENGAGE_ACCEPTANCE_SAHI' to value '40610'
12:08:30.447 [go] setting environment variable 'GO_FROM_REVISION_ENGAGE_ACCEPTANCE_SAHI' to value '40610'
12:08:30.447 [go] setting environment variable 'GO_DEPENDENCY_LABEL_ENGAGE_TRUNK' to value '2016.2.EAP.10027'
12:08:30.447 [go] setting environment variable 'GO_DEPENDENCY_LOCATOR_ENGAGE_TRUNK' to value 'engage-trunk/10027/masterdata/1'
12:08:30.447 [go] setting environment variable 'GO_REVISION_ENGAGEOPS' to value '40611'
12:08:30.447 [go] setting environment variable 'GO_TO_REVISION_ENGAGEOPS' to value '40611'
12:08:30.447 [go] setting environment variable 'GO_FROM_REVISION_ENGAGEOPS' to value '40611'
12:08:30.447 [go] setting environment variable 'nodename' to value 'satwehrlgaugerd'
12:08:30.447 [go] setting environment variable 'dbserver' to value 'satwehrsq12'
12:08:30.447 [go] setting environment variable 'AAT_SERVER' to value 'satwehrlgaugerd'
12:08:30.447 [go] setting environment variable 'AAT_EI_SERVER' to value '10.31.210.112:8082'
12:08:30.447 [go] setting environment variable 'nodename.ei' to value 'satwehrlaatsei'
12:08:30.447 [go] setting environment variable 'sourcedb' to value 'satwehrlgaugevanilla'
12:08:37.311 Note: Some input files use or override a deprecated API.
12:08:37.311 Note: Recompile with -Xlint:deprecation for details.
12:08:40.090 Executing in 2 parallel streams.
12:08:51.092 [runner: 2] Note: Some input files use or override a deprecated API.
12:08:51.092 [runner: 2] Note: Recompile with -Xlint:deprecation for details.
12:08:51.830 [runner: 1] Note: Some input files use or override a deprecated API.
12:08:51.830 [runner: 1] Note: Recompile with -Xlint:deprecation for details.
12:08:57.052 [runner: 2] Sahi properties file = C:\dev\sahi\config\sahi.properties[runner: 2] 
12:08:57.053 Sahi user properties file = E:\GoAgent\pipelines\engage-gauge-reference-data\engage.gauge\sahi\userdata\config\userdata.properties
12:08:57.221 [runner: 1] Sahi properties file = C:\dev\sahi\config\sahi.properties[runner: 1] 
12:08:57.221 Sahi user properties file = E:\GoAgent\pipelines\engage-gauge-reference-data\engage.gauge\sahi\userdata\config\userdata.properties
12:08:58.075 [runner: 2] >>>> Sahi OS v5.0 started. Listening on port: 9999[runner: 2] 
12:08:58.075 >>>> Configure your browser to use this server and port as its proxy
12:08:58.075 [runner: 2] >>>> Browse any page and CTRL-ALT-DblClick on the page to bring up the Sahi Controller
12:08:58.113 [runner: 2] -----[runner: 2] 
12:08:58.113 Reading browser types from: E:\GoAgent\pipelines\engage-gauge-reference-data\engage.gauge\sahi\userdata\config\browser_types.xml
12:08:58.122 [runner: 2] Opera was not found at C:\Program Files (x86)\Opera\opera.exe[runner: 2] 
12:08:58.122 Safari was not found at C:\Program Files (x86)\Safari\Safari.exe
12:08:58.122 Firefox 4 was not found at C:\Program Files (x86)\Mozilla Firefox 4.0 Beta 7\firefox.exe
12:08:58.122 -----
12:08:58.302 [runner: 2] Jul 13, 2016 12:08:58 PM net.sf.sahi.ssl.SSLHelper getKeyManagerFactoryForRemoteFetch
12:08:58.302 INFORMATION: No SSL Client Cert specified
12:08:58.886 [runner: 1] ---[runner: 1] 
12:08:58.886 --- ERROR: Port 9999 is already being used ---
12:08:58.886 ---
12:08:59.185 [runner: 2] Jul 13, 2016 12:08:59 PM net.sf.sahi.test.ProcessHelper execute
12:08:59.185 INFORMATION: "C:\Program Files (x86)\Google\Chrome\Application\chrome.exe" --incognito --user-data-dir=E:/GoAgent/pipelines/engage-gauge-reference-data/engage.gauge/sahi/userdata\browser\chrome\profiles\sahi0 --proxy-server=localhost:9999 --disable-popup-blocking "http://sahi.example.com/_s_/dyn/Driver_start?sahisid=sahi_ef4461520adc904e6b0aef508a32f6d95bb1__SahiAmpersandSahi__startUrl=http%3A%2F%2Fsahi.example.com%2F_s_%2Fdyn%2FDriver_initialized%3FstartUrl%3Dhttp%253A%252F%252Fsahi.example.com%252F_s_%252Fdyn%252FDriver_initialized"
12:09:00.422 [runner: 2] Keytool command found at: keytool[runner: 2] 
12:09:04.799 [runner: 2] # 000 Login to engage with Superadmin
12:09:04.800 [runner: 2] ## 000 Login to engage with Superadmin

the first runner is pointing out, that the port 9999 is already being used and so the tests cannot be run on another browser instance.

12:08:58.886 [runner: 1] ---[runner: 1] 
12:08:58.886 --- ERROR: Port 9999 is already being used ---

Is it possible to run the tests in parallel with sahi and if yes - how should I do it?

Thanks in advance.

Natalie

**Comment by kupna** [comment]
Is there any help you can give me to get this work?

**Comment by mahendrakariya** [comment]
Sorry for the delayed response. We haven't tried running specs using Sahi in parallel. But theoretically, this would involve initializing Sahi with a different proxy port for each parallel stream. @prateekbaheti may be able to shed some more light on this.

**Comment by kupna** [comment]
Thanks for the response. That's exactly what I want - but I don't know how to do it.

**Comment by kupna** [comment]
Any support on that?

**Comment by kupna** [comment]
@prateekbaheti could you please help me with that? I would really need that to run our tests faster. Thanks in advance.

**Comment by sriv** [comment]
@kupna - I have pushed a [commit](https://github.com/getgauge/gauge-example-sahi/commit/284a4edda08343597aae275e2c179b5e0a45b156) to the [gauge-example-sahi](https://github.com/getgauge/gauge-example-sahi) repository, which spins off sahi proxy in a new port thus enabling parallel run. Could you please pull the code and try running gauge specs in parallel?

**Comment by kupna** [comment]
@sriv I tried the new code but there is only port 9999 used for starting sahi with the tests. There is no IO Exception and so no other port number is tried. 

Also the browser window is not closed after executing the tests. Maybe because of that:
INFORMATION: PIDs not available yet. Waiting for 1 sec

see console output:

C:\dev\Gauge\bin\gauge.exe --simple-console --tags test --parallel C:/projects/engage/trunk/engage.gauge/specs
Executing in 2 parallel streams.
[runner: 2] Sahi properties file = C:\dev\sahi\config\sahi.properties[runner: 2] 
Sahi user properties file = C:\projects\engage\trunk\engage.gauge\sahi\userdata\config\userdata.properties
[runner: 1] Sahi properties file = C:\dev\sahi\config\sahi.properties[runner: 1] 
[runner: 1] Sahi user properties file = C:\projects\engage\trunk\engage.gauge\sahi\userdata\config\userdata.properties[runner: 1] 
[runner: 2] Starting SAHI proxy on port 9999[runner: 2] 
[runner: 1] Starting SAHI proxy on port 9999[runner: 1] 
[runner: 2] >>>> Sahi OS v5.0 started. Listening on port: 9999[runner: 2] 
>>>> Configure your browser to use this server and port as its proxy
>>>> Browse any page and CTRL-ALT-DblClick on the page to bring up the Sahi Controller
[runner: 2] -----[runner: 2] 
Reading browser types from: C:\projects\engage\trunk\engage.gauge\sahi\userdata\config\browser_types.xml
[runner: 2] Opera was not found at C:\Program Files (x86)\Opera\opera.exe[runner: 2] 
Safari was not found at C:\Program Files (x86)\Safari\Safari.exe
Firefox 4 was not found at C:\Program Files (x86)\Mozilla Firefox 4.0 Beta 7\firefox.exe
-----
[runner: 2] Aug 23, 2016 2:20:18 PM net.sf.sahi.ssl.SSLHelper getKeyManagerFactoryForRemoteFetch
INFORMATION: No SSL Client Cert specified
[runner: 1] ---[runner: 1] 
--- ERROR: Port 9999 is already being used ---[runner: 1] 
[runner: 1] ---[runner: 1] 
[runner: 2] Aug 23, 2016 2:20:19 PM net.sf.sahi.test.ProcessHelper execute
INFORMATION: "C:\Program Files (x86)\Google\Chrome\Application\chrome.exe" --args --new-window --incognito --user-data-dir=C:\projects\engage\trunk\engage.gauge\sahi\userdata/browser/chrome/profiles/sahi0 --proxy-server=localhost:9999 --disable-popup-blocking "http://sahi.example.com/_s_/dyn/Driver_start?sahisid=sahi_c7ffc9bc0fc73049050b27d049e14dff3e07__SahiAmpersandSahi__startUrl=http%3A%2F%2Fsahi.example.com%2F_s_%2Fdyn%2FDriver_initialized%3FstartUrl%3Dhttp%253A%252F%252Fsahi.example.com%252F_s_%252Fdyn%252FDriver_initialized"
[runner: 2] # test
[runner: 2]   ## test
[runner: 2] Aug 23, 2016 2:20:21 PM net.sf.sahi.test.ProcessHelper$PIDGatherer run
INFORMATION: PIDs: []; 1787 ms
[runner: 2] Aug 23, 2016 2:20:29 PM net.sf.sahi.RemoteRequestProcessor processHttp
WARNUNG: Returning CannotConnectResponse for: http://banoccevf/
[runner: 2] Aug 23, 2016 2:20:29 PM net.sf.sahi.RemoteRequestProcessor processHttp
WARNUNG: Returning CannotConnectResponse for: http://gdkhixbhtdshv/
[runner: 2] Aug 23, 2016 2:20:30 PM net.sf.sahi.RemoteRequestProcessor processHttp
WARNUNG: Returning CannotConnectResponse for: http://ikknbxkhotyszpp/
[runner: 2] 2016-08-23 14:20:35,035 1    [main] INFO  com.infoniqa.engage.login.LoginScreen  - User INFONIQA SUPERADMIN is logged IN.
[runner: 2] 
[runner: 2] # test1
[runner: 2]   ## test1
[runner: 2] 2016-08-23 14:21:06,378 31344 [main] INFO  com.infoniqa.engage.login.LoginScreen  - User INFONIQA SUPERADMIN is logged OUT.
[runner: 2] 2016-08-23 14:21:13,134 38100 [main] INFO  com.infoniqa.engage.login.LoginScreen  - User INFONIQA SUPERADMIN is logged IN.
[runner: 2] 2016-08-23 14:21:18,586 43552 [main] INFO  com.infoniqa.engage.login.LoginScreen  - User INFONIQA SUPERADMIN is logged OUT.
[runner: 2] 2016-08-23 14:21:25,946 50912 [main] INFO  com.infoniqa.engage.login.LoginScreen  - User INFONIQA SUPERADMIN is logged IN.
[runner: 2] 
[runner: 2] Aug 23, 2016 2:21:26 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:27 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:28 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:29 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:30 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:31 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:32 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:33 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:34 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:35 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:36 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:37 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:38 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:39 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:40 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:41 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:42 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:43 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:44 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:45 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:46 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:47 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:48 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:49 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:50 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:51 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:52 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:53 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:54 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:55 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:56 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:57 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:58 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:21:59 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:00 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:01 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:02 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:03 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:04 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:05 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:06 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:07 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:08 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:09 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:10 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec
[runner: 2] Aug 23, 2016 2:22:11 PM net.sf.sahi.test.ProcessHelper kill
INFORMATION: PIDs not available yet. Waiting for 1 sec

**Comment by kupna** [comment]
Please could you give me further help in this case?

**Comment by sriv** [comment]
@kupna - apologies for not answering earlier, got held up with elsewhere. Am looking at this right now.

**Comment by kupna** [comment]
@sriv thanks a lot :+1:

**Comment by sriv** [comment]
I just pushed [a commit](https://github.com/getgauge/gauge-example-sahi/commit/c7ee2d4cceb9f695219a82256ed64da8d3c78e3b) that randomizes that port for every runner. Can you please pull and check?

Also, I would like to point out that this is an example, and the real hack for getting sahi to start parallely is to use the code configuration and not depend on `browser_types.xml` configuration.

**Comment by kupna** [comment]
So I should remove the browser_types.cml configuration?

**Comment by sriv** [comment]
In the example, I deleted the `browser_types.xml` file from `sahi/userdata/config` and instead used the `Browser` constructor to [pass the options](https://github.com/getgauge/gauge-example-sahi/blob/master/src/test/java/com/thoughtworks/gauge/example/BrowserFactory.java#L54).

**Comment by kupna** [comment]
@sriv now it works that I get a free port for every runner, but after finishing the tests the browser windows are not closed as they are when I do not run the tests in parallel.

**Comment by kupna** [comment]
@sriv any input on that? thx

**Comment by sriv** [comment]
I am still trying to simulate a situation where the browsers would be orphaned. I can see that both firefox and chrome browser instances are getting killed as part of the `AfterSuite` hook. I am on mac osx (but I did try with ubuntu 14.04 as well), what operating system/browser versions are you seeing this issue?

My tests ran against Firefox 48.0.1, and chrome  52.0.2743.116.

**Comment by kupna** [comment]
I am running my tests on Windows 10 Enterprise with Chrome Version 52.0.2743.116

**Comment by kupna** [comment]
Sorry for the delay. I have been working on other problems in the meantime ;)

I already have the block look like the one you posted and it still does not work.

**Comment by kupna** [comment]
Any help on that?

**Comment by kupna** [comment]
could you please have another look at this problem? 

I still have the problem that when I run my tests in parallel the browser window is not closed. 

Additionally all processes are killed if one test of the tests running in parallel is ready.

**Comment by kupna** [comment]
@sriv can you help me?

**Comment by sriv** [comment]
@kupna - I will give this another go, and post my findings here.

**Comment by kupna** [comment]
@sriv thanks a lot :)

**Comment by sriv** [comment]
@kupna - just an update - I have setup a win10 box with chrome, and can see the stacktrace you mentioned.

This works for me on my macOS Sierra + Chrome. I will continue to investigate this.

**Comment by sriv** [comment]
Hi @kupna,

So, there seems to be something wrong when using this `Browser`:

```
                    browser = new Browser("$ProgramFiles (x86)\\Google\\Chrome\\Application\\chrome.exe",
                            "chrome.exe",
                            "--args --new-window --incognito --user-data-dir=" + userDataDir() + "\\browser\\chrome\\profiles\\sahi$threadNo --proxy-server=localhost:" + port + " --disable-popup-blocking",
                            LOCAL_HOST,
                            port);
```

For some reason, sometimes the proxy isn`t recognized, but it works some other times. But on windows, it never closed the browser for me.

I have tried a few more options:

Sahi has an option to reuse proxy and session for different browser instances, and setting `useSingleSession` to `true` lets you share the Sahi Session.

So I tried this:

- use `browser_types.xml` to configure browsers, and in the node for `chrome`, set `useSingleSession` = `true`.
- continue to use the same port and proxy.
- run `gauge -p specs`

The example code works, by sharing the same browser instance between runners. Now, I am unable to benchmark if this is actually a parallel run (and if this results in a gain in performance).

Can I request you to please try the above steps and see if it makes any difference in your case? Please let me know if you need any help in setting this up.

**Comment by kupna** [comment]
@sriv how does your BrowserFactory.java file look like, when you use it with browser_types.xml?

**Comment by kupna** [comment]
@sriv how do I get the runnerId into opening the browser when I open it with browser_types.xml?

**Comment by kupna** [comment]
@sriv can I have any help on that please?

**Comment by kashishm** [comment]
@kupna we tried a lot but we are unable to find the root cause. The code looks fine as `browser.close()` is being called in `AfterSuite` hook and Gauge is executing that hook. Please ask this question on Sahi's forum, you might get an answer there.

> The last option could be to write a script to kill the browsers after the specs execution.

**Comment by kupna** [comment]
@kashishm thats bad :( 

but how did @sriv try to run the processes in parallel as he mentioned before?

I wasn't able to try it, because I wasn't sure how he did it.

How does the browser_types.xml and the BrowserFactory.java file look for this try?

**Comment by sriv** [comment]
@kupna - there were two issues with my attempt:

- I was using a [build](https://github.com/getgauge/gauge-java/tree/multithreading) from a branch, and not master. This branch was created to investigate [another issue](https://github.com/getgauge/gauge#492). This was interfering with my investigation of this issue.

_[Sidenote: When getgauge/gauge#492 gets addressed, there is a chance that we can use the multithreading capability for this issue.]_

- When I built with master, and applied the changes I mentioned earlier, I noticed that the proxy and browser weren't always getting closed, leading us back to the previous issues.

In any case, here's what my code looks like:

`browser_types.xml`:

```
<browserTypes>
	<browserType> 
		<name>firefox</name> 
		<displayName>Firefox</displayName>
		<icon>firefox.png</icon>
		<path>open -n -a Firefox.app</path>  
		<options>--args -profile "$userDir/browser/ff/profiles/sahi$threadNo" -no-remote</options>  
		<processName>Firefox</processName> 
		<capacity>5</capacity>  
		<force>true</force>
	</browserType>
	
	<browserType> 
		<name>safari</name>  
		<displayName>Safari</displayName>
		<icon>safari.png</icon>
		<path>open -n -a Safari.app</path>  
		<options> </options>  
		<processName>Safari</processName> 
		<useSystemProxy>true</useSystemProxy>
		<capacity>5</capacity>
		<force>true</force>  
	</browserType>
	
	<browserType> 
		<name>chrome</name>  
		<displayName>Chrome</displayName>
		<icon>chrome.png</icon>
		<path>open -n -a "Google Chrome.app"</path>  
		<options>--args --incognito --user-data-dir=$userDir/browser/chrome/profiles/sahi$threadNo --proxy-server=localhost:9999 --disable-popup-blocking</options>  
		<processName>"Google Chrome"</processName> 
		<capacity>5</capacity>
		<useSingleSession>true</useSingleSession>
		<force>true</force>  
	</browserType>	
</browserTypes>
```

`BrowserFactory.java` looks very simple:

```
package com.thoughtworks.gauge.example;

import com.thoughtworks.gauge.AfterSuite;
import com.thoughtworks.gauge.BeforeSuite;
import net.sf.sahi.client.Browser;
import net.sf.sahi.config.Configuration;

import java.io.File;
import java.io.IOException;

public class BrowserFactory {

    private static Browser browser;

    @BeforeSuite
    public void setup() throws IOException {
        String sahiInstallDir = System.getenv("sahi.install_dir");
        if (sahiInstallDir == null || sahiInstallDir.isEmpty()) {
            throw new RuntimeException("Sahi Install directory not specified. Set in env/default/sahi_config.properties file");
        }
        Configuration.init(sahiInstallDir, userDataDir());
        launchBrowser();
    }

    @AfterSuite
    public void tearDown() {
        System.out.println("Cleaning up");
        if (browser != null) {
            System.out.println("Closing browser : " + browser.sessionId());
            browser.setSpeed(100);
            browser.close();
        }
    }

    private void launchBrowser() throws IOException {
        String browserName = System.getenv("sahi.browser_name");
        browser = new Browser(browserName);
        browser.open();
    }

    public static Browser getBrowser() {
        return browser;
    }

    private String userDataDir() {
        return new File("sahi", "userdata").getAbsolutePath();
    }
}
```

The premise of my theory is that, if you have a `Proxy` started, then newer instances of `Browser` should work independently, using the same proxy. This somewhat works, but I notice that the `Browser` loses its handle over the actual browser process and does not close it.

Apologies for having dragged this, and for the confusion this may have caused.

**Comment by kupna** [comment]
@sriv 
I saw, that you merged the multithreading branch of gauge and gauge-java into master, I thought, that I would give the parallel execution another try with the env enable_multithreading = true.

For now we have a linux machine to run the tests parallel, but we do not get it working.

It is really essentially for us, that we can run the tests in parallel, because when you run tests with a runtime of 20 minutes, you can imagine how long it takes to run them sequential, instead of parallel.

I would like to try to set up the test run completely like you do it, because at our environment everytime something is wrong.

Problems:

- it is not possible to assign different port numbers to run sahi on parallel
- the specs are running in parallel, but if one runner is ready the other runner will finish too even though it is not ready
- after running the tests the browser is not closed and the go agent cannot be used for other tests afterwards without manual killing of processes

So I need the exact instructions how to do this:

- how do the property files look like
- how does BrowserFactory.java look like?
- should I use browser_types.xml? (difficult with different sahi ports)

Did you try running the specs parallel with the gauge-example-sahi?

Please help me.

**Comment by Apoorva-GA** [comment]
@kupna We are looking into the issue and will keep you updated with our findings.

**Comment by kupna** [comment]
I have found some really essential things yesterday about parallel exectution of gauge tests with sahi.

BrowserFactory.java 
```java
package com.infoniqa.utils.sahi;

import com.thoughtworks.gauge.AfterSuite;
import com.thoughtworks.gauge.BeforeSuite;
import net.sf.sahi.Proxy;
import net.sf.sahi.client.BrowserCondition;
import net.sf.sahi.client.ExecutionException;
import net.sf.sahi.config.Configuration;
import org.apache.log4j.Logger;

import java.io.File;
import java.io.IOException;
import java.net.ServerSocket;
import java.nio.charset.Charset;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

public class BrowserFactory {

	private static CourseGuideBrowser browser;

	private static Logger logger = Logger.getLogger(BrowserFactory.class);
	public static final String LOCAL_HOST = "localhost";
	private static final int DEFAULT_PORT = 9999;
	private static Proxy proxy;

	@BeforeSuite
	public void setup() throws IOException {
		String sahiInstallDir = System.getenv("sahi.install_dir");
		if (sahiInstallDir == null || sahiInstallDir.isEmpty()) {
			throw new RuntimeException("Sahi Install directory not specified. Set in env/default/sahi_config.properties file");
		}
		Configuration.initJava(sahiInstallDir, userDataDir());
		launchBrowser();
	}

	@AfterSuite
	public void tearDown() throws IOException {
		removeSessionIdFile();

		if (browser != null) {
			browser.close();
		}
		if (proxy != null && proxy.isRunning()) {
			proxy.stop();
		}
	}

	private void launchBrowser() throws IOException {
		int port = portNumber();
		System.out.println("Starting SAHI proxy on port " + port);
		startProxy(port);
		String browserName = System.getenv("sahi.browser_name");
		int runnerId = DEFAULT_PORT - port;
		switch (browserName.toLowerCase()) {
			case "chrome":
				browser = new CourseGuideBrowser("$ProgramFiles (x86)\\Google\\Chrome\\Application\\chrome.exe",
						"chrome.exe",
						"--incognito --user-data-dir=$userDir\\browser\\chrome\\profiles\\sahi" + runnerId + " --proxy-server=localhost:" + port + " --disable-popup-blocking",
						LOCAL_HOST,
						port);
				break;
		}
		browser.open();

		createSessionIdFile();
	}

	private void removeSessionIdFile() {
		BrowserCondition condition = new BrowserCondition(browser) {
			@Override
			public boolean test() throws ExecutionException {
				Path file = getSessionIdFilePath();
				List<String> sessionIds = new ArrayList<>();
				try {
					sessionIds = Files.readAllLines(file, Charset.forName("UTF-8"));
					if (sessionIds.contains(browser.sessionId())) {
						sessionIds.remove(browser.sessionId());
						Files.write(file, sessionIds, Charset.forName("UTF-8"));
					}
					if (sessionIds.size() == 0) {
						Files.delete(file);
					}
				} catch (IOException e) {
					logger.error(e);
				}
				return sessionIds.size() == 0;
			}
		};
		browser.waitFor(condition, 30*60000); // 30 minutes
	}

	private void createSessionIdFile() throws IOException {
		Path file = getSessionIdFilePath();
		try {
			// Create the empty file with default permissions, etc.
			Files.createFile(file);
			logger.info(String.format("file created: %s",file));
		} catch (FileAlreadyExistsException x) {
			logger.info(String.format("file already exists: %s",file));
		} catch (IOException x) {
			// Some other sort of failure, such as permissions.
			logger.error("createFile error: %s%n", x);
		}
		List<String> sessionIds = Files.readAllLines(file, Charset.forName("UTF-8"));
		sessionIds.addAll(Arrays.asList(browser.sessionId()));
		Files.write(file, sessionIds, Charset.forName("UTF-8"));
	}

	private Path getSessionIdFilePath() {
		return Paths.get(userDataDir(), "sessionIds.txt");
	}

	private void startProxy(int port) {
		proxy = new Proxy(port);
		proxy.start(true);
		waitForProxy(10000);
	}

	public static CourseGuideBrowser getBrowser() {
		return browser;
	}

	private int portNumber() throws IOException {
		int port = ThreadLocalRandom.current().nextInt(9000, DEFAULT_PORT + 1);
		ServerSocket serverSocket = new ServerSocket(port);
		int freePort = serverSocket.getLocalPort();
		serverSocket.close();
		return freePort;
	}

	private String userDataDir() {
		return new File("sahi", "userdata").getAbsolutePath();
	}

	private void waitForProxy(int waitUntil) {
		int counter = 0;
		while (!proxy.isRunning() && counter < waitUntil) {
			try {
				Thread.sleep(1000);
				counter += 1000;
			} catch (InterruptedException e) {
				logger.error("wait for proxy failed", e);
			}
		}

		if (!proxy.isRunning()) {
			throw new RuntimeException("Sahi Proxy server is not starting");
		}
	}
}
```
First problem was, that my Browser creation was wrong - I used "chrome" instead of "chrome.exe" and so the net.sf.sahi.test.ProcessHelper could not find the new chrome processes and kill them after the run has finished.
So I changed it to

```java
	private void launchBrowser() throws IOException {
		int port = portNumber();
		System.out.println("Starting SAHI proxy on port " + port);
		startProxy(port);
		String browserName = System.getenv("sahi.browser_name");
		int runnerId = DEFAULT_PORT - port;
		switch (browserName.toLowerCase()) {
			case "chrome":
				browser = new CourseGuideBrowser("$ProgramFiles (x86)\\Google\\Chrome\\Application\\chrome.exe",
						"chrome.exe",
						"--incognito --user-data-dir=$userDir\\browser\\chrome\\profiles\\sahi" + runnerId + " --proxy-server=localhost:" + port + " --disable-popup-blocking",
						LOCAL_HOST,
						port);
				break;
		}
		browser.open();

		createSessionIdFile();
	}
```

The next problem I had after changing this, was that the processes are started in parallel but, if one process was ready, the other process was killed too, even though the process was not ready.

This is because in when you call browser.launch in net.sf.sahi.test.ProcessHelper there are all chrome.exe process ids stored and when you call browser.close then in net.sf.sahi.test.ProcessHelper all chrome.exe process ids are killed that have been created newly since browser.launch has been called. 

That is very inconvenient for running specs in parallel, because so every process with chrome.exe is killed even though the run is not ready.

net.sf.sahi.test.ProcessHelper
```java
 public void run() {
            ProcessHelper.this.status = ProcessHelper.Status.INITIALIZED;
            int maxTime = ProcessHelper.this.maxTimeToWaitForPIDs;
            int time = 0;

            for(byte interval = 100; !ProcessHelper.hasProcessStarted && time < maxTime; time += interval) {
                try {
                    Thread.sleep((long)interval);
                } catch (InterruptedException var6) {
                    var6.printStackTrace();
                }
            }

            try {
                ArrayList e = ProcessHelper.this.getAllPIDs();
                ProcessHelper.this.pids = ProcessHelper.this.getNewlyAdded(this.allPIDsBefore, e);
            } catch (Exception var5) {
                var5.printStackTrace();
            }

            if(!ProcessHelper.hasProcessStarted && time >= maxTime && ProcessHelper.this.retryCount++ < ProcessHelper.this.maxRetryCount) {
                try {
                    this.processHelper.kill();
                    ProcessHelper.this.releaseLock();
                    System.out.println(">>> ProcessHelper Retry Count: " + ProcessHelper.this.retryCount);
                    this.processHelper.execute();
                    return;
                } catch (Exception var7) {
                    var7.printStackTrace();
                }
            }

            ProcessHelper.logger.info("PIDs: " + ProcessHelper.this.pids + "; " + (System.currentTimeMillis() - ProcessHelper.t) + " ms");
            ProcessHelper.this.status = ProcessHelper.Status.RUNNING;
            ProcessHelper.this.releaseLock();
        }
```
Because I didn't want to change ProcessHelper I decided to store the browser.sessionIds in a file when browser.launch is called and delete the browser.sessionId out of this file when browser.close is called.

```java
private void createSessionIdFile() throws IOException {
		Path file = getSessionIdFilePath();
		try {
			// Create the empty file with default permissions, etc.
			Files.createFile(file);
			logger.info(String.format("file created: %s",file));
		} catch (FileAlreadyExistsException x) {
			logger.info(String.format("file already exists: %s",file));
		} catch (IOException x) {
			// Some other sort of failure, such as permissions.
			logger.error("createFile error: %s%n", x);
		}
		List<String> sessionIds = Files.readAllLines(file, Charset.forName("UTF-8"));
		sessionIds.addAll(Arrays.asList(browser.sessionId()));
		Files.write(file, sessionIds, Charset.forName("UTF-8"));
	}
```
```java
private void removeSessionIdFile() {
		BrowserCondition condition = new BrowserCondition(browser) {
			@Override
			public boolean test() throws ExecutionException {
				Path file = getSessionIdFilePath();
				List<String> sessionIds = new ArrayList<>();
				try {
					sessionIds = Files.readAllLines(file, Charset.forName("UTF-8"));
					if (sessionIds.contains(browser.sessionId())) {
						sessionIds.remove(browser.sessionId());
						Files.write(file, sessionIds, Charset.forName("UTF-8"));
					}
					if (sessionIds.size() == 0) {
						Files.delete(file);
					}
				} catch (IOException e) {
					logger.error(e);
				}
				return sessionIds.size() == 0;
			}
		};
		browser.waitFor(condition, 30*60000); // 30 minutes
	}
```
When I run it like this the browser waits to close until the file with the current browser.sessionIds is empty. And if browser.close is not colled, then the process ids with chrome.exe are not killed. 

But this is not really a parallel run, if I have to block the resources until the slowest one is ready. 

So what should I do to get this running real parallel? 

Is there any way to combine the parallel run of gauge with the parallel run of sahi?

Thanks for your help.

**Comment by kashishm** [comment]
@kupna We have made some changes to https://github.com/getgauge/gauge-example-sahi. Now, the example can be run using multithreading parallel execution and closes all the browsers at the end of execution.

Set `enable_multithreading` env/property to `true` to run in [multithreading parallel mode](https://docs.getgauge.io/using.html?highlight=parallel#parallel-execution-using-threads) and after that run `gauge -p specs/`.

> We tried it on mac and it is working as expected.

**Comment by Apoorva-GA** [comment]
Closing the issue. Let us know if you are still facing the issue.
